Make locations system field

jingcheng16 opened the pull request:
## Product Description
<!-- For non-invisible changes, describe user-facing effects. -->
Should have no visible changes, except for user cannot set `commcare_location_id`, `commcare_location_ids`, and `commcare_primary_case_sharing_id` in UserData directly 

## Technical Summary
<!--
    Provide a link to the ticket or document which prompted this change,
    Describe the rationale and design decisions.
-->
Ticket: https://dimagi.atlassian.net/browse/SAAS-15716

Please review by commit.


## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->
Write new tests for it.
This PR only changes where we pull data from, UserData or the `location_id` and `assigned_location_ids` attribute on CouchUser.

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->
`corehq.apps.user_importer.tests.test_importer:TestMobileUserBulkUpload` ensures location in user data will be updated when use bulk upload to update user's location.
New tests added in `corehq.apps.users.tests.test_user_data:TestUserDataModel` ensure location in user data will be updated when user's location changed.
### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->
QA Ticket: https://dimagi.atlassian.net/browse/QA-6781

QA plan:
- make changes to a mobile worker's location, and use User API to list all mobile workers, make sure what is returned in user data is expected. (If no location assigned, then no `location_id`, `location_ids` in user data)
- try to update `location_id` and `location_ids` in user data through User API and get error.



### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

jingcheng16 committed:
Make locations system field

jingcheng16 committed:
Fix AttributeError for missing location_ids attribute in CommCareUser

- Ensured that the system correctly handles cases where location information is not available in CommCareUser objects.

- Updated the _provided_by_system property to conditionally include COMMCARE_LOCATION_ID, COMMCARE_LOCATION_IDS and COMMCARE_PRIMARY_CASE_SHARING_ID.

jingcheng16 committed:
Add test for location change in user data

jingcheng16 committed:
Prevent modification of UserData with keys provided by the system

If location data (location_id, location_ids) for a CouchUser has never been set before, it won't be included in _provided_by_system. So I add _keys_provided_by_system

jingcheng16 committed:
Stop writing to user_data directly

jingcheng16 commented on corehq/apps/users/tests/test_user_data.py:
I added new tests first, and make sure previous implementation passed the new tests, then make changes to the code.

nospame commented on corehq/apps/users/tests/test_user_data.py:
```diff
@@ -430,3 +430,9 @@ def test_reset_locations_and_user_data_updated(self):
             'commcare_primary_case_sharing_id': self.loc1.location_id,
         })
+
+    def test_change_data_provided_by_system_will_raise_user_data_error(self):
```
nit: less verbosity along the lines of `test_changing_system_provided_data_raises_user_data_error`

nospame commented on corehq/apps/users/user_data.py:
```diff
@@ -174,8 +180,8 @@ def get(self, key, default=None):
     def __setitem__(self, key, value):
-        if key in self._provided_by_system:
-            if value == self._provided_by_system[key]:
+        if key in self._keys_provided_by_system:
+            if key in self._provided_by_system and value == self._provided_by_system[key]:
```
Maybe consider renaming `_keys_provided_by_system` to something more distinct from `_provided_by_system`, to make this easier to read, possibly `_system_keys`. For me, it was hard to figure out what these two lines actually mean without reading through them a few times. Not sure if you could also use `get` or `getattr` here but that's another approach that could help with clarity, along the lines of:
```python
if key in self._system_keys:
    if value == self._provided_by_system.get(key):
```
Unless we're concerned with someone trying to pass in a null value where one already exists, which would be interesting.

jingcheng16 replied:
Any value that is provided by system shouldn't be assigned in this set method. So even null value should be passed to it. But I will make changes to the function name!

jingcheng16 replied:
If we use `''` instead of `None` in getattr(self._couch_user, 'location_id', ''):, do we still need to user fallback like `object()` ?

esoergel replied:
Not sure, because setting it to `''` is also a real change, like `None`.  I'd expect both of these to fail:
```python
user_data['commcare_project'] = None
user_data['commcare_project'] = ''
```
Nice thing about `object()` is there's nothing you can pass in that will equal it.  You could also use a string like "MISSING", though that's a little messier.  `...` is also sometimes used for this sort of thing, as a placeholder meaning "not provided".

jingcheng16 replied:
Thank you! Fixed in: afdbd8d89c5a635e5214d4f2070027b42bbbf13e

biyeun commented:
I think it would be a good idea for @esoergel to review a lot of these changes, as he's most familiar with the locations code and can raise any issues that might have been missed

jingcheng16 commented on corehq/apps/users/tests/test_user_data.py:
```diff
@@ -289,3 +312,121 @@ def test_remove_unrecognized_empty_field(self):
+
+        # Set primary location to loc2
+        self.user.set_location(self.loc2)
+        self.assertEqual(user_data.to_dict(), {
```
I don't have a good answer. 😂 I just try, if it doesn't get updated, I refetch.

jingcheng16 replied:
Thanks @esoergel ... I still don't understand. If it is receiving the same `UserData` object, why we don't need to call `refresh_from_db` on that object?

esoergel replied:
It's a pass-by-reference.  `set_location` and this test both operate on not just the same user, but the same `user` instance and by extension the same `UserData` instance.  That is, if you called `id(user_data)` in both places, you'd get the same value.  This means it's functionally the same as doing:
```python
user_data = <get user data>
user_data['commcare_location_id'] = location.location_id
assertEqual(user_data.to_dict(), {...})
```
It doesn't matter that setting `commcare_location_id` happens in a different method, it's still mutating the same instance.

jingcheng16 replied:
I see!!! Thank you so much!!!

jingcheng16 commented on corehq/apps/users/user_data.py:
```diff
@@ -63,14 +64,19 @@ def save(self):
-            COMMCARE_LOCATION_IDS: self._couch_user.location_ids,
-            COMMCARE_PRIMARY_CASE_SHARING_ID: self._couch_user.location_id,
         }
+        if getattr(self._couch_user, 'location_id', None):
```
Here I want to keep the behavior remain unchanged. Before I make this change, if no location is assigned to the user, there won't be location info in user_data. The newly added test also ensures my changes won't change the content in user_data, it only changes where it comes from.

jingcheng16 replied:
Hi Ethan, yes, I would love to always include these fields. Can you point me to the code where we access the location info in user data. I'm just worried if there will be some check like: if `location_id` exist in `user_data`, then do something with the value of `location_id`. If the value is "", then that part of code might error.

esoergel replied:
Unfortunately that's what makes this so difficult to reason about - it isn't used in our code, it's used in app logic.  The restore file will contain something like
```xml
  <Registration xmlns="http://openrosa.org/user/registration">
    <username>ethan</username>
    <password>myhashedpassword</password>
    <uuid>myuserid</uuid>
    <date>2024-07-11</date>
    <user_data>
      <data key="commcare_first_name">Ethan</data>
      <data key="commcare_last_name">Sloergel</data>
      <data key="commcare_location_id">52f34e6469b240e6b8bdda7982df2cdd</data>
      <data key="commcare_location_ids">52f34e6469b240e6b8bdda7982df2cdd cab938fece104b888b558cc0b39e96d2</data>
      <data key="commcare_phone_number">9042410933</data>
      <data key="commcare_primary_case_sharing_id">52f34e6469b240e6b8bdda7982df2cdd</data>
      <data key="commcare_profile">2954</data>
      <data key="commcare_project">esoergel</data>
      <data key="commcare_user_type">commcare</data>
      <data key="favorite_color">purple</data>
      <data key="foo">baz</data>
      <data key="government_id"/>
      <data key="yearbook_quote">Not all who wander are lost</data>
    </user_data>
  </Registration>
```
That will be referenced in applications with expressions like
```
instance('commcaresession')/session/user/data/commcare_location_id
```
If that expression references a key that's not present, say `this_isnt_real`, it will raise an error like this:
```
Calculation Error: Error in calculation for /data/missing_field Logic references instance(commcaresession)/session/user/data/this_isnt_real which is not a valid question or value.
```
Whereas referencing a key that is present, but empty (like `government_id`) will return the empty string.

In practice, I think app builder often guard against this field being missing by doing something like this:
```
if(count(instance('commcaresession')/session/user/data/commcare_phone_number) > 0, instance('commcaresession')/session/user/data/commcare_phone_number, "missing")
```
That conditional will behave differently if a field is empty than if it's missing.  That probably means this change is too risky to be warranted, though it is a frustrating situation for everyone, especially app builders.

jingcheng16 replied:
Thank you! Yeah I remember we talked about it when we're having the call. That's why I want to make sure I didn't change any behavior. Btw, I cannot find Matt's PR that saving all user fields, can you give me a link? If we all agree it is risky, can we say we don't need to change my current implementation?

esoergel replied:
Matt's PR is here: https://github.com/dimagi/commcare-hq/pull/34082
Yeah I think it's appropriate to not change behavior here

nospame commented on corehq/apps/users/tests/test_user_data.py:
```diff
@@ -289,3 +312,121 @@ def test_remove_unrecognized_empty_field(self):
+            'commcare_primary_case_sharing_id': self.loc1.location_id,
+        })
+
+    def test_reset_locations_and_user_data_updated(self):
```
I see that line of reasoning. Since it's testing the effect of the action on the user data, what would you think about phrasing it as `updates_user_data` instead? To me, that reads much more clearly, like "performing x action has y result", where the result is that it updates `user_data`.

jingcheng16 replied:
`udpates_user_data` sounds more natural! Thank you! Updated!

jingcheng16 committed:
REFACTOR: rename test and functions

esoergel commented on corehq/apps/users/models.py:
```diff
@@ -2099,7 +2097,6 @@ def set_location(self, location, commit=True):
             raise AssertionError("You can't set an unsaved location")
 
         user_data = self.get_user_data(self.domain)
-        user_data['commcare_location_id'] = location.location_id
```
beautiful stuff

jingcheng16 commented on corehq/apps/users/tests/test_user_data.py:
```diff
@@ -140,8 +143,18 @@ def _get_profile(self, profile_id):
 
 
 @patch('corehq.apps.users.user_data.UserData._get_profile', new=_get_profile)
-class TestUserDataModel(SimpleTestCase):
```
Yes I also find `test_location_assignment` covers it!  I think it is no harm to add more test? Should I remove the new tests or move it to `TestUserData` ?

jingcheng16 replied:
Sure I'll remove them except for `test_change_data_provided_by_system_will_raise_user_data_error` also rename it to `test_changing_system_provided_data_raises_user_data_error` as suggested by Evan.

jingcheng16 commented on corehq/apps/users/tests/test_user_data.py:
```diff
@@ -289,3 +312,121 @@ def test_remove_unrecognized_empty_field(self):
+        self.assertEqual(user_data.to_dict(), {
+            'commcare_project': self.domain,
+            'commcare_profile': '',
+        })
```
Agree! Will make those changes once I know if I should move my tests to `TestUserData` or should I remove it?

esoergel commented on corehq/apps/users/user_data.py:
```diff
@@ -63,14 +64,19 @@ def save(self):
-            COMMCARE_LOCATION_IDS: self._couch_user.location_ids,
-            COMMCARE_PRIMARY_CASE_SHARING_ID: self._couch_user.location_id,
         }
+        if getattr(self._couch_user, 'location_id', None):
```
It occurs to me that this is a difference between web and mobile workers - for web, the fields are always present, but for mobile, they're only present if non-empty.  That is a gap - I didn't consider the significance of that distinction when implementing the web users bit, but I think we should bring it in to line.  If that makes sense to do as part of this PR, great, otherwise I can follow up with it afterwards.

jingcheng16 replied:
Hi Ethan, what do you mean by "bring it in to line"? I saw `CouchUser` has `get_location_id` , and both `CommCareUser` and `WebUser` have `get_location_ids`. I will just call this two methods instead of access the fields directly. What else do you want me to include in this PR?

esoergel replied:
That should be correct, thank you.  I mean that right now web users always have `commcare_location_id` in their restores, even when not assigned a location, because of that first block of code I linked above.  This I now think is an inconsistency, and I'd like to give web users the same behavior as mobile workers.  Happy to do that in another PR though.

jingcheng16 replied:
Thank you Ethan!
- Will make `comcare_location_id` only present if non-empty have an unexpected impact for web user?
- "web users don't have a assigned_location_ids attribute" > I see we don't store `location_id` and `assigned_location_ids` for `WebUser`, but we can still access them on `WebUser` because they are attributes for `CouchUser`, it is just empty. Would move them to `CommCareUser` make sense?
- I made changes in 911410a3d3ad80abb721be05449dc52ecd255fc1 and 5193c764639da7e6468172de0c06db4d3278d53d. But haven't fix the inconsistency. I would prefer having that fix in a separate PR and make sure it is safe to make that change.

jingcheng16 replied:
Ethan, my change failed on `test_user_data_ignores_location_fields` (https://github.com/dimagi/commcare-hq/actions/runs/9912924492/job/27388787564?pr=34852). Seems like previous behavior is, WebUser don't have `commcare_location_id` and `commcare_location_ids` in their user data! Now after my change, it will has... I'm concerned of change the existing behavior. Also I don't know if I by accident achieved the "bring it in to line"... I might still not fully understand what you mean and what you're trying to achieve. 😞

jingcheng16 replied:
Ethan, my change failed on `test_user_data_ignores_location_fields` (https://github.com/dimagi/commcare-hq/actions/runs/9912924492/job/27388787564?pr=34852). Seems like previous behavior is, WebUser don't have `commcare_location_id` and `commcare_location_ids` in their user data! Now after my change, it will has... I'm concerned of change the existing behavior. Also I don't know if I by accident achieved the "bring it in to line"... I might still not fully understand what you mean and what you're trying to achieve. 😞

jingcheng16 replied:
Ethan, my change failed on `test_user_data_ignores_location_fields` (https://github.com/dimagi/commcare-hq/actions/runs/9912924492/job/27388787564?pr=34852). Seems like previous behavior is, WebUser don't have `commcare_location_id` and `commcare_location_ids` in their user data! Now after my change, it will has... I'm concerned of change the existing behavior. Also I don't know if I by accident achieved the "bring it in to line"... I might still not fully understand what you mean and what you're trying to achieve. 😞

esoergel replied:
Yeah, for web users, the location fields are only available in the restore and user cases, not in user data directly.  I think for this PR it should be sufficient to add a check to only insert them in user data for mobile workers, but I'd still go through the `get_location_id` type accessors.

jingcheng16 committed:
Refactor: use get and fall back to object()

fall back to object() so nothing will equal it if key doesn't exist, better than fall back to None or ''

jingcheng16 committed:
use get_location_id and get_location_ids

jingcheng16 committed:
use get_location_id and get_location_ids

esoergel commented on corehq/apps/users/user_data.py:
```diff
@@ -60,12 +67,34 @@ def save(self):
+                provided_data[COMMCARE_LOCATION_IDS] = user_location_data(
+                    self._couch_user.get_location_ids(self.domain))
+
+        if settings.UNIT_TESTING:
+            # Some test don't have an actual user existed
+            if self._couch_user:
+                _add_location_data()
+        else:
+            _add_location_data()
```
nit: you can combine these conditionals and avoid needing to pull the body out like so:
```suggestion
        # Some test don't have an actual user existed
        if self._couch_user or not settings.UNIT_TESTING:
            _add_location_data()
```

jingcheng16 committed:
Web User don't store location fields in user data

jingcheng16 committed:
Add decorator back and add comments

The decorator is missed by accident in https://github.com/dimagi/commcare-hq/pull/34224 so I'm adding it back

jingcheng16 committed:
Remove tests that are already covered by other tests

we already have a good portion of these tests in
https://github.com/dimagi/commcare-hq/blob/1a70089d74f4c7c367799ef83e7eddcd2739340a/corehq/apps/users/tests/test_location_assignment.py

jingcheng16 commented:
@esoergel Hi Ethan, do you mind giving a final round of review?
I removed most tests I added, but kept `test_no_location_info_in_user_data_when_no_location_assigned` and `test_change_data_provided_by_system_will_raise_user_data_error`
I added comment to `test_user_data_ignores_location_fields` because when it failed and find it confusing. I also realized test in `TestWebUserBulkUpload` might have some issue, because when I remove `@patch('corehq.apps.user_importer.importer.domain_has_privilege', lambda x, y: True)`, test `test_web_user_location_add`, `test_web_user_location_remove` and `test_invite_location_add` won't fail. I expect it to fail because in `importer.py`, we check if `self.domain_info.can_assign_locations` before we update locations. But I don't plan to fix this in my PR, as I have a tight timeline for User API work.

esoergel commented on corehq/apps/user_importer/tests/test_importer.py:
```diff
@@ -1839,7 +1839,12 @@ def test_user_data_profile_removal(self):
     def test_uncategorized_data(self):
         self._test_uncategorized_data(is_web_upload=True)
 
+    @patch('corehq.apps.user_importer.importer.domain_has_privilege', lambda x, y: True)
```
There's also a `@privilege_enabled` test decorator that you might be interested in

esoergel commented on corehq/apps/users/user_data.py:
```diff
@@ -60,12 +67,32 @@ def save(self):
+                provided_data[COMMCARE_LOCATION_IDS] = user_location_data(
+                    self._couch_user.get_location_ids(self.domain))
+
+        # Some test don't have an actual user existed
+        # Web User don't store location fields in user data
+        if self._couch_user and self._couch_user.is_commcare_user() or not settings.UNIT_TESTING:
+            _add_location_data()
```
I don't think this conditional is quite right, if the user is a web user, this conditional will be: `if True and False or True`, which evaluates to `True`, meaning we'll attempt to add location data.
```suggestion
        # Some test don't have an actual user existed
        # Web User don't store location fields in user data
        if (self._couch_user or not settings.UNIT_TESTING) and self._couch_user.is_commcare_user():
            _add_location_data()
```

jingcheng16 committed:
Fix conditional logic

Co-authored-by: Ethan Soergel <esoergel@users.noreply.github.com>

jingcheng16 committed:
Remove decorator as the domain already have all the privileges

jingcheng16 commented:
QA find out, when we remove all location from one mobile worker, the api call will still return those old `location_id` and `location_ids` of that mobile worker. Because when remove all locations, the mobile worker's `location_id` and `assigned_location_id` would be empty, then it won't be returned in `_provided_by_system` to overwrite the existing location fields in `UserData`. I'll follow up with the migration to delete all location fields in UserData for mobile workers, (because web user don't store location fields in their user data). I'll run migration manually after the PR is merged. ( I'll run migration for india and swiss on Wednesday afternoon, and for prod on Thursday morning.)

FYI @esoergel

jingcheng16 committed:
Revert "Stop writing to user_data directly"

This reverts commit bee93c56e57e5f34bb209cdc8e1b99d9a9d830b5.

jingcheng16 committed:
Reapply "Stop writing to user_data directly"

This reverts commit e30a5dfd5d65d8c830fcef74090a149fdff02ea3.

jingcheng16 committed:
Temporary code to make sure we don't read from existing user data is there is nothing in _provided_by_system

esoergel commented on corehq/apps/users/user_data.py:
```diff
@@ -60,19 +67,45 @@ def save(self):
+        # Ensure location keys are only from _provided_by_system
+        for key in LOCATION_KEYS:
+            if key in combined_dict and key not in self._provided_by_system:
+                del combined_dict[key]
```
The idea is that system fields shouldn't come from `_local_to_user`, so it'd make sense to me to remove them only from there, rather than the combined dict, something like this:
```python
**{k: v for k, v in local_to_user.items() if k not in self._system_keys}
```
You could do that here, or even in `__init__`, which would also mean these would be removed from other places, like `raw` and the places that modify `_local_to_user`.  It'd also mean that these fields would be removed from the DB whenever user data models are saved (which might be helpful with the migration).

As-is, I think doing `del user_data['commcare_location_id']` won't raise an error, though it probably should.

jingcheng16 replied:
Hi Ethan, I love you advice and implement in bdba44ea47df35f2fde857e43d40ce325d3b66fa. However I didn't add it in __init__, I think it is an overkill, as we stop writing location info into UserData in hq, and also forbid this behavior in `__set__`, `update`.

jingcheng16 committed:
Prevent any update for _system_keys

jingcheng16 commented on corehq/apps/users/user_data.py:
```diff
@@ -162,8 +195,8 @@ def get(self, key, default=None):
     def __setitem__(self, key, value):
-        if key in self._provided_by_system:
-            if value == self._provided_by_system[key]:
+        if key in self._system_keys:
```
Hi Ethan, that's my oversight. I thought no location fields will be in userdata after migration, so we will never encounter situation that people want to update or delete or pop location fields in userdata. But I think if I consider the time after the code is deployed and before the migration, I should use `_system_keys` in all places. Fixed in e28058fc34c3ddfcb119fc66756bebb56201b15a

jingcheng16 committed:
exclude system keys from _local_to_user

jingcheng16 commented:
I've reverted [this migration](https://github.com/dimagi/commcare-hq/pull/34899) on staging, so I can test the new commit I made to this PR. [QA-6846](https://dimagi.atlassian.net/browse/QA-6846) still pass.
